doc/neps/nep-0052-python-api-cleanup.rst

```diff
@@ -122,53 +147,77 @@ probably. Here is a representative set of examples:
   splitting out like was done for ``np.financial``).
-- The ``np.compat`` namespace can be removed.
-- There are lots of one-off functions that can be deprecated and removed, such as
-  ``real_if_close`` (see `gh-11375 <https://github.com/numpy/numpy/issues/11375>`__).
-  These can be identified via triaging the issue tracker and/or via going
-  through the main namespace manually.
 
-There are new namespaces for warnings/exceptions (``np.exceptions``) and for dtype-related
-functionality (``np.types``). NumPy 2.0 is a good opportunity to move things
-there from the main namespace.
+S: I think `np.financial` maintainer was interested in adopting?
+
+- The ``np.compat`` namespace, used during the Python 2 to 3 transition, will be removed.
+- Functions that are narrow in scope, with very few public use-cases,
+  will be removed.  See, e.g.
+  ``real_if_close`` (`gh-11375 <https://github.com/numpy/numpy/issues/11375>`__).
+  These will have to be identified manually and by issue triage.
+
+New namespaces are introduced for warnings/exceptions (``np.exceptions``) and for dtype-related
+functionality (``np.types``). NumPy 2.0 is a good opportunity to
+popular these submodules from the main namespace.
 
-Functionality that gets a fair amount of usage but has a preferred alternative
+.. S: Has the ``np.types`` name been fixed? Wonder if we're going to
+   create confusion with that name.
+
+Functionality that is widely used but has a preferred alternative
 may be hidden by not including it in ``__dir__``, rather than deprecating it. A
 ``.. legacy::`` directory may be used to mark such functionality in the
 documentation.
 
-A more comprehensive test will be added to the test suite to ensure that we won't get
-any more accidental additions to any namespace - every new entry will need to be
-allow-listed.
+.. S: Perhaps we should give ourselves more freedom here: delete
+   unused functions, deprecate common functions for which an
+   alternative is recommended. We can raise a very helpful exception
+   in these cases, telling the developer exactly what to change.
+
+A test will be added to ensure limited future growth of all
+namespaces; i.e., every new entry will need to be explicitly added to
+an allow-list.
 
 Cleaning up the submodule structure
 ```````````````````````````````````
 
+.. S: This conversation is relevant:
+   `MAINT: Hide internals of np.lib to only show submodules <https://github.com/numpy/numpy/pull/18447>`__.
+   It looked at the time that we were all roughly on board, just that
+   we could not justify such a big change.
+
 Let's reorganize the API reference guide along main and submodule namespaces,
 and only within the main namespace use the current subdivision along
 functionality groupings. Also by "mainstream" and special-purpose namespaces.
-Details TBD, something like::
+Details TBD, something like:
+
+.. S: not sure what to call these submodules; made something up, but
+   should be improved
```

Comment:
I'm not sure about this, I did not intend to move existing submodules that we want to keep to new names like `util` or `algorithms`. That would break a lot of stuff. I was thinking keeping the names as is, but grouping them like I described in the docs. Because our reference guide is quite badly structured right now.

Reply:
Great, that resolves the confusion on my part; I couldn't figure out how that restructure would work. Leaving it as is makes more sense. Maybe NumPy 3.0 :)

Reply:
Great, glad we cleared that up.